Add keypress & output control services to Envisalink component

jnimmo opened the pull request:
**Description:**
Added two new services to the alarm_control_panel required for Envisalink functionality

alarm_keypress allows a string of custom keypresses to be sent to the alarm 
alarm_output_control activates the specified programmable output on the alarm

**Pull request in [home-assistant.io](https://github.com/home-assistant/home-assistant.io) with documentation (if applicable):** home-assistant/home-assistant.io#1249

**Checklist:**

If user exposed functionality or configuration variables are added/changed:
- [x] Documentation added/updated in [home-assistant.io](https://github.com/home-assistant/home-assistant.io)

If the code communicates with devices, web services, or third-party tools:
- [x] Local tests with `tox` run successfully. **Your PR cannot be merged unless tests pass**
- [x] New dependencies have been added to the `REQUIREMENTS` variable ([example](https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/keyboard.py#L16)).
- [x] New dependencies are only imported inside functions that use them ([example](https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/keyboard.py#L51)).
- [x] New dependencies have been added to `requirements_all.txt` by running `script/gen_requirements_all.py`.

jnimmo committed:
Add keypress & output control services to Envisalink component

Add services to allow sending custom keypresses and activating
programmable outputs on an alarm control panel.
Implemented for the Envisalink alarm, and moving to new version of
pyenvisalink to support this.

Replicated the service handler mapping code from Cover component into
Alarm Control Panel to allow handling alternative schemas if required
by new services.

jnimmo committed:
Add keypress & output control services to Envisalink component

Add services to allow sending custom keypresses and activating
programmable outputs on an alarm control panel.
Implemented for the Envisalink alarm, and moving to new version of
pyenvisalink to support this.

Replicated the service handler mapping code from Cover component into
Alarm Control Panel to allow handling alternative schemas if required
by new services.

jnimmo committed:
Update requirements_all.txt

jnimmo committed:
Update requirements_all.txt

jnimmo committed:
Updated services.yaml

jnimmo committed:
Updated services.yaml

jnimmo commented on homeassistant/components/alarm_control_panel/__init__.py:
```diff
@@ -146,6 +182,14 @@ def alarm_trigger(self, code=None):
 
+    def alarm_keypress(self, keypress=None):
+        """Send custom key sequence to alarm."""
+        pass
```
Thanks for your review - Just on recommendation in pull request 3884 where this originally started - so the other alarm services aren't required to implement it

balloob replied:
Any method that is exposed as a service should raise `NotImplementedError()`

balloob commented:
Any new service for components need also be added to the demo platform and tests need to be written.

However I am not convinced that these services are going to be implemented for any other platform so I would like to suggest to move them inside the envisalink platform and register them there. [Sonos](https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/media_player/sonos.py) does a similar thing.

jnimmo committed:
Removed requirement to enter code in HA UI

Incorporated changes suggested by @sriram
https://github.com/srirams/home-assistant/commit/2f8deb70cb5f3621a69b6b9
acb72f8e29123650c

Including pending state for exit/entry delay

Clarified services to use the code passed to them as a first priority,
otherwise use the code from configuration

Swapped back to using NotImplementedError for the service definitions

jnimmo committed:
Removed requirement to enter code in HA UI

Incorporated changes suggested by @sriram
https://github.com/srirams/home-assistant/commit/2f8deb70cb5f3621a69b6b9
acb72f8e29123650c

Including pending state for exit/entry delay

Clarified services to use the code passed to them as a first priority,
otherwise use the code from configuration

Swapped back to using NotImplementedError for the service definitions

jnimmo commented:
So is that a no balloob unless I move the services within Envisalink? It is pretty standard functionality for wired alarms to have programmable outputs, i.e. on the DSC alarms you activate PGM2 if you need to reset the smoke detector; and PGM1 might be controlling the flashing light on the bell box; but it was my initial concern that putting it in the main class might be rejected unless other alarm control panels were implemented at the same time.
Is the manual alarm just used for testing and demo platform? if so I can add tests to that

balloob commented:
I must admit that I am not very familiar with wired alarms. Can someone else chime in if this is common in alarms?

(if it goes into the main component, you'll need to add it to the demo component and include tests for it.)

persandstrom commented:
Sorry,  I don't know much about wired alarms. The ALARM_OUTPUT_CONTROL_SCHEMA looks specific even if the functionallity might be common.

jnimmo committed:
- Add support for alarm_keypress to manual alarm (functions like a standard alarm keypad where entering the code disarms or arms the alarm)
- Add tests for alarm_keypress to manual alarm
- Style corrections (too many returns, comment & whitespace issues)

jnimmo committed:
- Add support for alarm_keypress to manual alarm (functions like a standard alarm keypad where entering the code disarms or arms the alarm)
- Add tests for alarm_keypress to manual alarm
- Style corrections (too many returns, comment & whitespace issues)

jnimmo commented:
I've added support to the manual alarm for alarm_keypress so it can be tested and manual alarm is used in the demo platform.

Alarm_output_control is trickier as has no practical application for a virtual alarm panel, and can't really be tested on Envisalink as it is just an interface to a device?

I can always remove the alarm_output_control service and just implement alarm_keypress if you think that is more widely applicable (because the outputs can be controlled with keypresses anyway). I would prefer not to register the services inside Envisalink as that complicates the component and duplicates service processing code..

jnimmo committed:
Removed alarm_output_control service as unable to incorporate in the demo/test in a meaningful way

jnimmo committed:
Removed alarm_output_control service as unable to incorporate in the demo/test in a meaningful way

jnimmo commented:
Hi @balloob, I've just removed the alarm_output_control service as I can see that has more limited uses than the keypress service. Could you please let me know if there are any other changes required. This also upgrades to the latest pyenvisalink library which adds support for last_armed_by and last_disarmed_by users which members of the forum are having to manually update to currently. Thanks

balloob commented:
Adding a generic service is easy. Getting rid of it is hard and having to support it in every UI for Home Assistant ever is also a pain. So let's just keep it inside envisalink for now until other platforms will show that there is a generic need.

jnimmo committed:
Removed alarm_keypress from alarm_control_panel

# Conflicts:
#	homeassistant/components/alarm_control_panel/__init__.py

jnimmo committed:
Merge branch 'home-assistant/dev' into alarm-keypress-handler

jnimmo committed:
Moved the Alarm_Keypress service into Envisalink component out of the generic

jnimmo commented:
Thanks for the reply - no problem I've figured out how to move the service into the component. Should be ready to pull now, Cheers

balloob committed:
Update envisalink.py

balloob committed:
Update services.yaml

balloob commented:
Bueno 🐬
